### ISSUE TITLE
TOPPView crashes

### ISSUE THREAD

**Issue by cbielow** [OpenMS/OpenMS#5318]
due to new functionality added in #5256 

when closing TOPPView, the app crashes:

```
Exception thrown at 0x00007FFEE7E04F38 in TOPPView.exe: Microsoft C++ exception: std::bad_alloc at memory location 0x000000DA16EFDC90.
Debug Error!
abort() has been called

(Press Retry to debug the application)
TOPPView.exe has triggered a breakpoint.
```
with the following stacktrace:

```
OpenMS_GUId.dll!std::_Associated_state<OpenMS::Param>::_Get_value(bool _Get_only_once) Line 285
	at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\include\future(285)
OpenMS_GUId.dll!std::_Task_async_state<OpenMS::Param>::_Get_value(bool _Get_only_once) Line 705
	at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\include\future(705)
OpenMS_GUId.dll!std::_State_manager<OpenMS::Param>::_Get_value() Line 787
	at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\include\future(787)
OpenMS_GUId.dll!std::future<OpenMS::Param>::get() Line 894
	at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\include\future(894)
OpenMS_GUId.dll!OpenMS::TVToolDiscovery::waitForParams::__l2::<lambda>() Line 85
	at C:\dev\openms_test\src\openms_gui\source\VISUAL\TVToolDiscovery.cpp(85)
OpenMS_GUId.dll!OpenMS::TVToolDiscovery::waitForParams() Line 72
	at C:\dev\openms_test\src\openms_gui\source\VISUAL\TVToolDiscovery.cpp(72)
OpenMS_GUId.dll!OpenMS::TOPPViewBase::addToolParamsToIni() Line 1669
	at C:\dev\openms_test\src\openms_gui\source\VISUAL\APPLICATIONS\TOPPViewBase.cpp(1669)
OpenMS_GUId.dll!OpenMS::TOPPViewBase::savePreferences() Line 1650
	at C:\dev\openms_test\src\openms_gui\source\VISUAL\APPLICATIONS\TOPPViewBase.cpp(1650)
OpenMS_GUId.dll!OpenMS::TOPPViewBase::~TOPPViewBase() Line 2872
	at C:\dev\openms_test\src\openms_gui\source\VISUAL\APPLICATIONS\TOPPViewBase.cpp(2872)
TOPPView.exe!main(int argc, const char * * argv) Line 187
	at C:\dev\openms_test\src\openms_gui\source\VISUAL\APPLICATIONS\GUITOOLS\TOPPView.cpp(187)

```

I'm not sure that parameters should be created when closing down toppview...

The reason it crashes, it that I have not build all TOPP tools/ UTILS in my development version.
Normally, this would lead to a Messagebox saying "Error: the Tool XY could not be found". But I guess showing this messagebox while TOPPViewbase is in teardown is a bad idea....

So basically, this behaviour might not be optimal. On first glance i'm not even sure why preferences are saved when TOPPView is closed (they should be saved then closing the preferences dialog...).
Can someone investigate this?

**Comment by timosachsenberg** [comment]
@dvdvgt any idea how to e.g. cancel the parsing ?

**Comment by dvdvgt** [comment]
I'm on it.

> On first glance i'm not even sure why preferences are saved when TOPPView is closed (they should be saved then closing the preferences dialog...).

Well this is nothing new that was added in my PR. Previously the preferences were also saved to the .ini when the TOPPViewBase deconstructor was called. 

However I did not consider the case that some tools might not be build. There's probably no quick way fix this. I'll try the following:

* `ToolsDialog` assumes that a param for each tool/util listed in ToolHandler is available. I'll try that it just throws a warning that the tool could not be loaded if its selected.
* I'll have to make sure that `TVToolDiscovery::getParamFromIni_` does not crash when a tool executable was not found and maybe also throw a warning.

**Comment by cbielow** [comment]
> Well this is nothing new that was added in my PR. Previously the preferences were also saved to the .ini when the TOPPViewBase deconstructor was called.

True. But maybe its enough to just not do that (unless there is a reason why this should be done, that I'm not aware of; --- ahh... maybe its the recently used files which need updating... -- but if that is the case, this can be done when a new file is opened. Writing the param object is really fast and it has the advantage that if TOPPView crashes, the file list would not get lost if the d'tor is not called).

> * ToolsDialog assumes that a param for each tool/util listed in ToolHandler is available. I'll try that it just throws a warning that the tool could not be loaded if its selected.
> * I'll have to make sure that TVToolDiscovery::getParamFromIni_ does not crash when a tool executable was not found and maybe also throw a warning.

Sounds good!

Also, are the parameters currently queried when TOPPView starts? Because I did not see any crashes at either startup nor when openning+closing the preferences (even though some TOPP tools were missing). So I'm not sure the TVToolDiscovery ran before....

Sorry for being vague here. I did not verify anything in the codebase; just making wild guesses here.

**Comment by dvdvgt** [comment]
> ToolsDialog assumes that a param for each tool/util listed in ToolHandler is available. I'll try that it just throws a warning that the tool could not be loaded if its selected.

I stand corrected. `ToolsDialog` is fine with missing params since it checks each tool param and only adds those that are compatible with the currently selected layer type. When an empty param at `TVToolDiscovery::getParamFromIni_` is returned the corresponding tool will not be added to the tool list in ToolsDialog.

> I'll have to make sure that TVToolDiscovery::getParamFromIni_ does not crash when a tool executable was not found and maybe also throw a warning.

This might actually do it. I tested it by removing some tool executables and running `TOPPView --force` so that the params are queried at startup. You'll just get some errors written to stderr that the tools could not be loaded. Exiting TOPPView also seems to be happening gracefully now without a crash. It seems that saving the tool params in `TOPPViewBase`s deconstructor to `TOPPView.ini` also posses no issues anymore as tool params that failed to be loaded in `TVToolDiscovery::getParamFromIni_` only manifest as empty params which are just not written to the .ini.

Can you try replacing `TVToolDiscovery::getParamFromIni_` with the following code (just the try catch block is different from the current version) and check whether the crashes on exit persist?

```cpp
Param TVToolDiscovery::getParamFromIni_(const std::string &tool_name)
  {
    // Temporary file path and arguments
    String path = File::getTemporaryFile();
    String working_dir = path.prefix(path.find_last_of('/'));
    QStringList args{"-write_ini", path.toQString()};
    // Start tool/util to write the ini file
    Param tool_param;
    String executable;
    // Return empty param if tool executable cannot be found
    try
    {
      executable = File::findSiblingTOPPExecutable(tool_name);
    }
    catch (const Exception::FileNotFound& e)
    {
      std::cerr <<  e << std::endl;
      return tool_param;
    }

    ExternalProcess proc;
    auto return_state = proc.run(executable.toQString(), args, working_dir.toQString(), false, ExternalProcess::IO_MODE::NO_IO);
    // Return empty param if writing the ini file failed
    if (return_state != ExternalProcess::RETURNSTATE::SUCCESS)
    {
      return tool_param;
    }
    // Parse ini file to param object and return it
    ParamXMLFile paramFile;
    paramFile.load((path).c_str(), tool_param);
    return tool_param;
  }
```

> Also, are the parameters currently queried when TOPPView starts? 

There are three different possibilities described in `TOPPViewBase::TOOL_SCAN`:
1. `SKIP_SCAN`: The tool params are not queried unless `ToolsDialog` is opened where the params are needed. This option is not exposed publicly and can only be used internally for e.g. testing.
2. `SCAN_IF_NEWER_VERSION`: The tool params are only queried at startup when the OpenMS version and the version in `TOPPView.ini` differ or the `TOPPView.ini` has no tool params i.e. no `params` section. This is the standard behavior. 
3. `FORCE_SCAN`: The tool params are always queried at startup. Can be triggered by starting toppview with `TOPPView --force`.

**Comment by timosachsenberg** [comment]
Not related to this issue: @dvdvgt did you also check what happens if the users closes TOPPView while tools are scanned?

**Comment by dvdvgt** [comment]
> Not related to this issue: @dvdvgt did you also check what happens if the users closes TOPPView while tools are scanned?  

Yes, it's fine. The deconstructor of TOPPViewBase waits until the scan has finished. Since the scan only takes about a second it shouldn't be noticeable. Of course if the TOPPView process is killed forcefully (e.g. ctrl+c) that's undefined behaviour.

**Comment by hendrikweisser** [comment]
Looks like a fix has been merged (#5335). Closing this issue.
